Enable focus events in cmdline and terminal modes

expipiplus1 opened the pull request:
This change adds switch cases for K_FOCUSGAINED and K_FOCUSLOST to the
input handling functions in ex_getln.c and terminal.c. The handling is
identical to what's found in edit.c (just calling apply_autocmds).

Fixes #3714

expipiplus1 committed:
Enable focus events in cmdline and terminal modes

This change adds switch cases for K_FOCUSGAINED and K_FOCUSLOST to the
input handling functions in ex_getln.c and terminal.c. The handling is
identical to what's found in edit.c (just calling apply_autocmds).

If one enters cmdline-mode by feeding `:` and sends a focuslost event (by
leaving the window for example) the text `<FocusLost>` will be inserted
into the command line. There is similar behaviour in terminal mode. This
patch corrects this behavior to fire the apropriate autocmd instead.

Fixes #3714

mhinz commented:
The problem:

Without this patch you would literal key inserts in your cmdline, terminal etc. when switching focus. E.g. `:`, switch back and forth, and you'll get `<FocusLost><FocusGained>` which is obviously not what you want.

:arrow_right: LGTM

Pyrohh commented:
> Without this patch you will get literal key inserts in your cmdline, terminal etc. when switching focus. E.g. :, switch window back and forth, and you'll get :<FocusLost><FocusGained>, which is obviously not what you want.

This, or the text of https://github.com/neovim/neovim/issues/3714, should probably be in the commit message.

Pyrohh commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -150,8 +151,11 @@ describe('tui', function()
   end)
 
   it('can handle focus events', function()
+    execute('set noshowmode')
```
I might be missing something obvious, but why is this needed?

expipiplus1 replied:
If this line isn't here then in insert mode the cmdline becomes `--INSERT--` instead of displaying any messages (such as `gained` or `lost`).

Echoing messages seemed like the simplest way of testing the events firing.

Pyrohh commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -173,6 +177,79 @@ describe('tui', function()
+                                                        |
+      lost                                              |
+      -- TERMINAL --                                    |
+    ]])
```
Since these tests are all related to perhaps these should be refactored into individual `it()` tests inside a `describe()` block?

expipiplus1 committed:
Reorganize focus events test into individual tests

The focus event tests now live in their own `describe` block with each
test testing the handling of focus events in a single mode.

expipiplus1 commented:
@Pyrohh I've reorganized the tests into their own describe block.

mhinz commented:
Merged. Thanks!
